lib/prometheus-parser/src/lib.rs

```diff
@@ -364,13 +644,187 @@ mod test {
+                    }, )* ],
+                    samples: vec![
+                        $( proto::Sample { value: $sample as f64, timestamp: $timestamp as i64 }, )*
+                    ],
+                }, )* ],
+            }
+        };
+    }
+
+    #[test]
+    fn parse_request_empty() {
+        let parsed = parse_request(write_request!([], [])).unwrap();
+        assert!(parsed.is_empty());
+    }
+
+    #[test]
+    fn parse_request_only_metadata() {
+        let parsed = parse_request(write_request!(["one" = Counter, "two" = Gauge], [])).unwrap();
+        assert_eq!(parsed.len(), 2);
+        match_group!(parsed[0], "one", Counter => |metrics: &MetricMap<SimpleMetric>| {
+            assert!(metrics.is_empty());
+        });
+        match_group!(parsed[1], "two", Gauge => |metrics: &MetricMap<SimpleMetric>| {
+            assert!(metrics.is_empty());
+        });
+    }
+
+    #[test]
+    fn parse_request_untyped() {
+        let parsed = parse_request(write_request!(
+            [],
+            [ [__name__ => "one", big => "small"] => [123 @ 1395066367500] ]
+        ))
+        .unwrap();
+
+        assert_eq!(parsed.len(), 1);
+        match_group!(parsed[0], "one", Untyped => |metrics: &MetricMap<SimpleMetric>| {
+            assert_eq!(metrics.len(), 1);
+            assert_eq!(
+                metrics.get_index(0).unwrap(),
+                simple_metric!(Some(1395066367500), labels!(big => "small"), 123.0)
+            );
+        });
+    }
+
+    #[test]
+    fn parse_request_gauge() {
+        let parsed = parse_request(write_request!(
+            ["one" = Gauge],
+            [
+                [__name__ => "one"] => [ 12 @ 1395066367600, 14 @ 1395066367800 ],
+                [__name__ => "two"] => [ 13 @ 1395066367700 ]
+            ]
+        ))
+        .unwrap();
+
+        assert_eq!(parsed.len(), 2);
+        match_group!(parsed[0], "one", Gauge => |metrics: &MetricMap<SimpleMetric>| {
+            assert_eq!(metrics.len(), 2);
+            assert_eq!(
+                metrics.get_index(0).unwrap(),
+                simple_metric!(Some(1395066367600), labels!(), 12.0)
+            );
+            assert_eq!(
+                metrics.get_index(1).unwrap(),
+                simple_metric!(Some(1395066367800), labels!(), 14.0)
+            );
+        });
+        match_group!(parsed[1], "two", Untyped => |metrics: &MetricMap<SimpleMetric>| {
+            assert_eq!(metrics.len(), 1);
+            assert_eq!(
+                metrics.get_index(0).unwrap(),
+                simple_metric!(Some(1395066367700), labels!(), 13.0)
+            );
+        });
+    }
+
+    #[test]
+    fn parse_request_histogram() {
+        let parsed = parse_request(write_request!(
+            ["one" = Histogram],
+            [
+                [__name__ => "one_bucket", le => "1"] => [ 15 @ 1395066367700 ],
+                [__name__ => "one_bucket", le => "+Inf"] => [ 19 @ 1395066367700 ],
```

Comment:
What does `le` stand for? I found it used it https://prometheus.io/docs/practices/histograms, but still couldn't figure it out 😄

Reply:
"less than or equal to" Each bucket's value is the sum of itself and all the ones smaller than it.